wallet/src/main/scala/org/bitcoins/wallet/sync/WalletSync.scala

```diff
@@ -0,0 +1,85 @@
+package org.bitcoins.wallet.sync
+
+import org.bitcoins.core.protocol.blockchain.{Block, BlockHeader}
+import org.bitcoins.core.util.{BitcoinSLogger, FutureUtil}
+import org.bitcoins.crypto.DoubleSha256DigestBE
+import org.bitcoins.wallet.Wallet
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait WalletSync extends BitcoinSLogger {
+
+  def sync(
+      wallet: Wallet,
+      getBlockHeaderFunc: DoubleSha256DigestBE => Future[BlockHeader],
+      getBestBlockHashFunc: () => Future[DoubleSha256DigestBE],
+      getBlock: DoubleSha256DigestBE => Future[Block])(implicit
+      ec: ExecutionContext): Future[Wallet] = {
+    val bestBlockHashF = getBestBlockHashFunc()
+    val bestBlockHeaderF = for {
+      bestBlockHash <- bestBlockHashF
+      bestheader <- getBlockHeaderFunc(bestBlockHash)
+    } yield bestheader
+
+    val blocksToSyncF = for {
+      bestHeader <- bestBlockHeaderF
+      blocksToSync <- getBlocksToSync(wallet = wallet,
+                                      currentTipBlockHashBE = bestHeader.hashBE,
+                                      accum = Vector.empty,
+                                      getBlock = getBlock)
+    } yield blocksToSync
+
+    val syncedWalletF = for {
+      blocksToSync <- blocksToSyncF
+      syncedWallet <- FutureUtil.foldLeftAsync(wallet, blocksToSync) {
+        case (wallet, nextBlock) =>
+          wallet.processBlock(nextBlock)
+      }
+    } yield syncedWallet
+
+    syncedWalletF
+  }
+
+  /** Syncs the wallet by walking backwards from the currentTip until we reach our wallet's best blockHash */
+  private def getBlocksToSync(
+      wallet: Wallet,
+      currentTipBlockHashBE: DoubleSha256DigestBE,
+      accum: Vector[Block],
+      getBlock: DoubleSha256DigestBE => Future[Block])(implicit
+      ec: ExecutionContext): Future[Vector[Block]] = {
+    val initSyncDescriptorOptF = wallet.getSyncDescriptorOpt()
+    for {
+      syncDescriptorOpt <- initSyncDescriptorOptF
+      walletBestHash = syncDescriptorOpt match {
+        case Some(descriptor) => descriptor.bestHash
+        case None             => wallet.chainParams.genesisHashBE
+      }
+      currentBlockOpt <- {
+        if (walletBestHash == currentTipBlockHashBE) {
+          Future.successful(None) // done syncing!
+        } else {
+          getBlock(currentTipBlockHashBE)
+            .map(Some(_))
+        }
+      }
+      blocks <- {
+        currentBlockOpt match {
+          case Some(currentBlock) =>
+            //loop again as we need to keep syncing
+            getBlocksToSync(wallet = wallet,
+                            currentTipBlockHashBE =
+                              currentBlock.blockHeader.previousBlockHashBE,
```

Comment:
should validate that the prev block isn't genesis, otherwise this function could run forever

Reply:
Done on line 60